File: COMP60611/source/C-labs/lab5/sblas_1.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <omp.h>
#include<mpi.h>

void vector_update(double a[], double b[], double x[], int limit);
void matrix_update(double **l, double y[], int limit);
void vector_add(double a[], double b[], double x[], int limit);
double vec_sum(double x[], int limit); 
void trimv(double **l, double y[], double a[], int limit); 

/* SBLAS code */

void *main() {

    int n = 1024*60;
    int m = 512;
    int nsteps = 1000;
    int k, i;
    MPI_Status status;

    //double ll[m][m], **l;
    double **l;
    double a[n], b[n], x[n], y[m];
    //double x2[n];
    double res;
    double vector_start, sum_start, trimv_start, vec_update_start, matrix_update_start, total_start;

    double vec_update_time = 0.0;
    double vector_time = 0.0;
    double sum_time = 0.0;
    double matrix_update_time = 0.0;
    double trimv_time = 0.0;
    double total_time = 0.0;    
    double ires;
    // MPI specific data
    int myid, numprocs, rc, ierr;
    int nx;

    // MPI initialisation
    printf("hello world\n");
    ierr = MPI_Init(NULL, NULL); 
    MPI_Comm_rank(MPI_COMM_WORLD, &myid);
    MPI_Comm_size(MPI_COMM_WORLD, &numprocs);
    printf ("Process %d of %d is alive\n", myid, numprocs);
  
    // allocate (lower triangular) matrix as a contibuous block and set up pointers to allow l[][] access 
    l = (double **) malloc((unsigned) m*sizeof(double*));  
    for (i=0; i <= m-1; i++) 
    {
        l[i] = (double *)malloc((unsigned) m * sizeof(double));
        //l[i] = ll[i];
    }

    // Loop over whole program to get reasonable execution time
    // (simulates a time-steping code)

    if (myid == 0) total_start = omp_get_wtime();

    for (k=1; k<=nsteps; k++ )  
    {
        //vec_update_start = omp_get_wtime();
        vector_start = omp_get_wtime();
        if (myid == 0) vector_update(a, b, x, n);
        //vec_update_time = vec_update_time + (omp_get_wtime() - vec_update_start);

        // Work out size of data on this processor (not a general solution!)

        nx = n/numprocs;

        // Distribute the vectors a and b to other processors

        //if (myid == 0) 
        //{
        //    // MASTER
        //    // send parts of the vectors a and b to other processors (x not reqd)

        //    for (int i=1; i<numprocs; i++) 
        //    {
	    //        ierr = MPI_Send(&a[i*nx], nx, MPI_DOUBLE_PRECISION, i, 0, MPI_COMM_WORLD);
	    //        ierr = MPI_Send(&b[i*nx], nx, MPI_DOUBLE_PRECISION, i, 0, MPI_COMM_WORLD);
        //    }
        //} 
        //else 
        //{
        //    // WORKER
        //    // receive data
        //    ierr = MPI_Recv(a, nx, MPI_DOUBLE_PRECISION, 0, 0, MPI_COMM_WORLD, &status);
        //    ierr = MPI_Recv(b, nx, MPI_DOUBLE_PRECISION, 0, 0, MPI_COMM_WORLD, &status);
        //}

        // vector addition operation: each processor adds its own parts 
        // Note: only master time is collected...

        // vector addition operation
        //vector_start = omp_get_wtime();
        MPI_Scatter(a, nx, MPI_DOUBLE_PRECISION,a,nx,MPI_DOUBLE_PRECISION,0,MPI_COMM_WORLD);
        MPI_Scatter(b, nx, MPI_DOUBLE_PRECISION,b,nx,MPI_DOUBLE_PRECISION,0,MPI_COMM_WORLD);

        //vector_start = omp_get_wtime();
        vector_add(a, b, x, nx);
        //vector_time = vector_time + (omp_get_wtime() - vector_start);

        
        MPI_Gather(x, nx, MPI_DOUBLE_PRECISION,x,nx,MPI_DOUBLE_PRECISION,0,MPI_COMM_WORLD);
        // Gather result vector on processor 0

        //if (myid == 0) 
        //{
        //    // MASTER
        //    // receive parts of result into the appropriate parts of my x

        //    for (i=1;i<numprocs; i++) 
        //    {
	    //        ierr = MPI_Recv((x+i*nx), nx, MPI_DOUBLE_PRECISION,	\
		//	    i, 0, MPI_COMM_WORLD,			\
		//	    &status);
        //    }
    
        //} 
        //else 
        //{
        //    // WORKER
        //    // send my part of result vector
        //    ierr = MPI_Send(x, nx, MPI_DOUBLE_PRECISION,	\
		//      0, 0, MPI_COMM_WORLD);
        //}

        // Rest of the code is performed on processor 0

        vector_time = vector_time + (omp_get_wtime() - vector_start);
        MPI_Barrier(MPI_COMM_WORLD);
        
        if (myid == 0) 
        {

            // sum operation

            sum_start = omp_get_wtime();
        }
        res = vec_sum(x,nx);
        MPI_Reduce(&res, &ires, 1, MPI_DOUBLE_PRECISION, MPI_SUM, 0, MPI_COMM_WORLD);
        //res = vec_sum(x2,nx);
        //ires = vec_sum(x, n);
        
        if (myid == 0)
        {
            sum_time = sum_time +(omp_get_wtime() - sum_start);

            // 'Update' lower triangular matrix and y vector

            matrix_update_start = omp_get_wtime();
            matrix_update(l, y, m);
            matrix_update_time = matrix_update_time +(omp_get_wtime() - matrix_update_start);

            // triangular matrix times vector operation

            trimv_start = omp_get_wtime();
            trimv(l, y, a, m);
            trimv_time = trimv_time + (omp_get_wtime() - trimv_start);

        }

        // end time-step loop
    }
  
    if (myid == 0)  total_time = omp_get_wtime() - total_start;
  
    for (i=0;i<numprocs;i++) 
    {
        MPI_Barrier(MPI_COMM_WORLD);
        if (myid == i) 
        {
            if (myid == 0) 
            { //MASTER
	            printf("--MASTER output\n");
	            //printf( "vector update time = %f seconds \n \n", vec_update_time);
	            printf( "--x[0] = %f \n", x[0]);
	            printf( "--x[n-1] = %f \n", x[n-1]);
	            printf( "--vector add time = %f seconds \n \n", vector_time);
	            printf( "--Sum of x() = %f \n", ires);
	            printf( "--sum time = %f \n \n", sum_time);
	            printf( "--matrix update time = %f seconds \n \n", matrix_update_time);
	            printf( "--y[0] = %f \n", y[0]);
	            printf( "--y[m-1] = %f \n", y[m-1]);
	            printf( "--trimv time = %f seconds \n \n", trimv_time);
	            printf( "--Total time = %f seconds \n \n", total_time);
	            printf("--End MASTER output\n");
            }
            printf("myid=%d\n",myid);
            printf( "vector add time = %f seconds \n \n", vector_time);
        }
    }

    MPI_Finalize();
}

void vector_update(double a[], double b[], double x[] ,int limit) {
  int i;

  for (i=0; i<limit; i++ )  {
    x[i] = 0.0;
    a[i] = 3.142+i;
    b[i] = 3.142;
  }
}

void matrix_update(double **l, double y[], int limit) {
  int i, j;

  for (i=0; i<limit; i++ ) {
    y[i] = 0.0;
    for (j=0; j<=i; j++ ) { 
      l[i][j] = 2.0;
    }
  }
}

void vector_add(double a[], double b[], double x[], int limit) {
  int i;

  for (i=0; i<limit; i++ ) {
    x[i] = a[i] + b[i];
  }
}

double vec_sum(double x[], int limit) {
  int i;
  double sum = 0.0;

  for (i=0; i<limit; i++ ) {
    sum = sum + x[i];
  }
  return sum;
}

void trimv(double **l, double y[], double a[], int limit) {
  int i, j;

  for (i=0; i<limit; i++) {
    for (j=0; j<=i; j++) { 
      y[i] = y[i] + l[i][j]*a[j];
    }
  }
}

File: COMP60611/source/lab5/sblas_1.f90
program sblas
    
  use mpi
  use timers
  type(TimerType) :: vecaddTime,vecsumTime,trimvTime,totalTime
  
  integer i,j,n,m,k,nsteps
  integer nx
  integer status(MPI_STATUS_SIZE)
  parameter (n=1024*60,m=512)
  parameter (nsteps=1000)
  real*8 a(n),b(n),x(n),y(m),l(m,m)
  real*8 buff(n)
  common /heap/ a,b,x,y,l
  real*8 res

! MPI specific data

  integer myid, numprocs, rc, ierr

! ..Intrinsic functions

! .. Executable statements

! MPI initialisation

  call MPI_INIT( ierr )
  call MPI_COMM_RANK( MPI_COMM_WORLD, myid, ierr )
  call MPI_COMM_SIZE( MPI_COMM_WORLD, numprocs, ierr )
  print *, "Process ", myid, " of ", numprocs, " is alive"

  call TimerCreate(vecaddTime,"vecadd")
  call TimerCreate(vecsumTime,"vecsum")
  call TimerCreate(trimvTime,"trimv")
  if (myid.eq.0) call TimerCreate(totalTime,"total")

! Loop over whole program to get reasonable execution time
! (simulates a time-steping code)

! start total timer on master (proc 0)

  if (myid.eq.0) call TimerOn(totalTime)

  do k=1,nsteps

! 'Update' vector data on processor 0

     if (myid.eq.0) call vector_update(n,x,a,b)

! Work out size of data on this processor (not a general solution!)

     nx = n/numprocs

! Distribute the vectors a and b to other processors

     if (myid.eq.0) then
! MASTER
! send parts of the vectors a and b to other processors (x not reqd)

        do i=1,numprocs-1
           call MPI_SEND(a(i*nx+1), nx, MPI_DOUBLE_PRECISION, & 
                I, 0, MPI_COMM_WORLD, ierr)
           call MPI_SEND(b(i*nx+1), nx, MPI_DOUBLE_PRECISION, &
                I, 0, MPI_COMM_WORLD, ierr)
        end do
     else
! SLAVE
! receive data

        call MPI_RECV(a, nx, MPI_DOUBLE_PRECISION, &
             0, 0, MPI_COMM_WORLD, &
             status, ierr)
         call MPI_RECV(b, nx, MPI_DOUBLE_PRECISION, &
              0, 0, MPI_COMM_WORLD, &
              status, ierr)
      end if

! vector addition operation: each processor adds its own parts 
! Note: only master time is collected...

      call TimerOn(vecaddTime)
      call vecadd (nx,x,a,b)
      call TimerOff(vecaddTime)

! Gather result vector on processor 0

      if (myid.eq.0) then
! MASTER
! receive parts of result into the appropriate parts of my x

         do i=1,numprocs-1
            call MPI_RECV(x(i*nx+1), nx, MPI_DOUBLE_PRECISION, &
                 i, 0, MPI_COMM_WORLD, &
                 status, ierr)
         end do

      else
! SLAVE
! send my part of result vector

         call MPI_SEND(x, nx, MPI_DOUBLE_PRECISION, &
              0, 0, MPI_COMM_WORLD, ierr)

      end if

! Rest of the code is performed on processor 0

      if (myid.eq.0) then

! sum operation

         call TimerOn(vecsumTime)
         call vecsum(n,res,x)
         call TimerOff(vecsumTime)

! 'Update' lower triangular matrix and y vector

         call matrix_update(m,y,l)

! triangular matrix times vector operation

         call TimerOn(trimvTime)
         call trimv (m,y,l,a)
         call TimerOff(trimvTime)

      end if

! end time-step loop
   end do

   if (myid == 0) call TimerOff(totalTime)
! print out some results

   do i=1,numprocs
     call MPI_Barrier(MPI_COMM_WORLD, ierr)
     if (myid == (i-1)) then
       if (myid.eq.0) then
          print *,'MASTER output'
          print *,'x(1) = ',x(1),'x(n) = ', x(n)
          print *,'Sum of x() = ',res
          print *,'y(1) = ',y(1),'y(m) = ', y(m)
          call TimerPrint(totalTime)
          call TimerPrint(vecsumTime)
          call TimerPrint(trimvTime)
          print *,'End MASTER output'
          call flush()
! end of processor 0 processing
       end if

       print *,'myid= ', myid
       call TimerPrint(vecaddTime)
       call flush()
     end if
   end do
   call MPI_FINALIZE(ierr)
 end program sblas

! Subroutines

 subroutine vector_update(n,x,a,b)

! .. Scalar arguments
   integer n

! .. Array arguments
   real*8 x(n),a(n),b(n)

! .. Local scalars
   integer i

! .. External functions

   do i=1,n
      x(i) = 0.
      a(i) = 3.142
      b(i) = 3.142
   end do

   return
 end subroutine vector_update

 subroutine matrix_update(m,y,l)

! .. Scalar arguments
   integer m

! .. Array arguments
   real*8 y(m),l(m,m)

! .. Local scalars
   integer i,j

! .. External functions

   do i=1,m
      y(i) = 0.
      do j=1,i
         l(j,i) = 2.
      end do
   end do

   return
 end subroutine matrix_update

 subroutine vecadd (n,x,a,b)

! .. Scalar arguments
   integer n

! .. Array arguments
   real*8 x(n),a(n),b(n)

! .. Local scalars
   integer i

! .. External functions

   do i=1,n
      x(i) = a(i) + b(i)
   end do
   return
 end subroutine vecadd

 subroutine vecsum(n,sum,a)

! .. Scalar arguments 
   integer n
   real*8 sum

! .. Array arguments
   real*8 a(n)

! .. Local scalars
   integer i

   sum = 0.0
   do i=1,n
      sum = sum +a(i)
   end do
   return
 end subroutine vecsum

 subroutine trimv (m,y,l,a)

! .. Scalar arguments
   integer m

! .. Array arguments
   real*8 y(m),l(m,m),a(m)

! .. Local scalars
   integer i,j

! .. External functions

   do i=1,m
      do j=1,i
         y(i) = y(i) + l(j,i)*a(j)
      end do
   end do
   return
 end subroutine trimv

